OpenRA.Game/Server/Server.cs

```diff
@@ -885,6 +1048,9 @@ public void DropClient(Connection toDrop)
 				// Send disconnected order, even if still in the lobby
 				DispatchOrdersToClients(toDrop, 0, Order.FromTargetString("Disconnected", "", true).Serialize());
 
+				if (!dropClient.IsObserver)
+					SetPlayerDefeat(toDrop.PlayerIndex);
```

Comment:
That is why I suggested we move this kind of logic into the game. Ending the match when all opposition is defeated or disconnected shouldn't be too hard and not as controversial as surrendering on disconnect.

Reply:
IMO its not going to be that simple: if *the server* gets disconnected then there will be no clients left to decide the end state, and we are no better off than what we have now.

Reply:
I don't see how you can get anything if the server gets disconnected.

Reply:
That was my point: we can't avoid all cases where the server needs to write out unresolved replays, so changing the end game logic handling doesn't change anything regarding the replay feature.

Reply:
It removes the need to count disconnected players as defeated by a special case for server replays here.

Reply:
My previous point was that a special case shouldn't be needed even with our current end-game logic 😉

Reply:
Which is fair enough. For the record I'm not arguing against adding disconnect timestamps.

Reply:
> I don't think this is correct. The game client doesn't consider disconnected players as defeated until all of their stuff has been destroyed or captured.
> 
> Consider the situation where Player A has Player B on the ropes, but drops from the game 2 seconds before their units destroy the final enemy building. 2 seconds later, the (auto)targeting finishes the job, and Player B is defeated. The client replay will show that A defeated B, but the server replay will show that B defeated A... right?

Maybe, but I think the client replay would be wrong then. I'm treating the Disconnect as a Surrender.

> IMO, going even further, this logic shouldn't be in the replay saving at all because this is applying interpretation to the game results, which other usecases may disagree with. So IMO the best solution would be to keep the replay metadata recording the undetermined result, have the player metadata record the tick/time that each player disconnected, and then your replay parsing tool can have a check to say "if state is undetermined, take the player/team who disconnected last as the winner"

This makes the whole thing much more complex for the replay parser: while extracting the game info is pretty straightforward, simple and fast, demuxing the whole replay requires an order of magnitude of additional complexity on my side. If this chunk is not accepted, I'll likely run a modified version of the OpenRA server to save me all that trouble of parsing the whole replay.

I think it's fairly innocent to have that information in the replay server, especially since the defeat/victory system is pretty broken currently for client replays.

Reply:
I removed the 2 lines of disconnect handling from the PR, and will make another when this one is merged.

Reply:
Emphasis on this part:
> have the player metadata record the tick/time that each player disconnected

I'm suggesting that we store this alongside the rest of the game/player metadata so that you won't need to parse the order stream :)